Move React ppx from compiler repo, add tests

jchavarri opened the pull request:
I find myself wanting to contribute to React ppx but always end up giving up because of the lack of infrastructure. By moving the file here, we can leverage the testing infrastructure and keeping it more in sync with the parser.

If this makes sense and gets merged, I will update the compiler repo to read from new file.

Hopefully this change makes sense as well from organizational perspective, as the React ppx is a syntax transformation.

Summary of changes:
- Copied the React ppx file from:
https://github.com/rescript-lang/rescript-compiler/blob/4f4812aa4e1411029392f6dbcacc580e755d68d5/jscomp/syntax/reactjs_jsx_ppx.cppo.ml, and then applied these changes:
  - Add `@raise` annotations
  - Remove dead functions
  - Removed macro `#ifdef BINARY`
  - Remove unused [`Ast_mapper.register` call](https://github.com/rescript-lang/rescript-compiler/blob/4f4812aa4e1411029392f6dbcacc580e755d68d5/jscomp/syntax/reactjs_jsx_ppx.cppo.ml#L972-L974)
  - Format file using width 120
- Add `reactjs_jsx_ppx.mli` file
- Added `-reactJsx` option to `rescript.exe` (for now I didn't added it to command `-help` output as I'm not sure if it should be exposed)
- Adapt test runner to interpret test files `.react.res` and `react.resi`
- Add a few tests

Thanks to these changes, I already found 2 bugs:
- Comments at the top of the file are placed in the wrong position
- `<Foo.bar />` elements are not supported by new parser

jchavarri commented on tests/runner.js:
```diff
@@ -173,7 +189,13 @@ function printFile(filename) {
   args.push(filename);
 
-  return cp.spawnSync(parser, args).stdout.toString("utf8");
+  let result = cp.spawnSync(parser, args);
+  
+  return {
+    result: result.stdout.toString("utf8"),
+    status: result.status,
+    errorOutput: result.stderr
+  }
```
I'm not sure about this. The idea was to give an informative error when a test file `react.res` is failing to parse, because otherwise tests would pass and an empty string will be stored as snapshot.

But maybe it's not necessary.

IwanKaramazow commented:
@jchavarri This is great! Will take a look at your PR. Long term we want to merge the jsx ppx with the parser itself, i.e. do the syntax transform in the parser.

jchavarri commented:
Thanks for the responsiveness @IwanKaramazow! Much appreciated :) 

> Long term we want to merge the jsx ppx with the parser itself.

I realized while working on this that there is no need to have it as a separate step now. It makes sense to do so, although maybe in mid-long term? e.g. 6-12 months. I mention that because I guess that merging the ppx with the parser would potentially break conversion between code in Reason and ReScript syntaxes? 🤔 Which is really useful while this transitions happens.

Also, I don't know if it's possible, but it'd be great to keep it as a separate module, so that tools like https://jchavarri.github.io/ppx-explorer/ that rely on a PPX-like API can keep updates with upstream easily, e.g [here](https://github.com/jchavarri/ppx-explorer/blob/525d5fccdae7e61bee3b5d903a604ae7a4251f3c/src/worker/RefmtWorker.re#L23).

chenglou commented:
Hey Javier! It's been a while
We're gonna pause this specifically, because the React PPX might change in the future (cc @rickyvetter). Switching version without breakage would require a big syntax-level switch, which would need some consideration.

jchavarri commented:
@chenglou hey! long time indeed :)

Could you elaborate, please? This PR doesn't change the ppx semantics, and it does not contain any breaking changes in the ppx code. All changes in `Reactjs_jsx_ppx` module are purely cosmetic (formatting, add reanalyze `@raise` attributes etc).

If there is some breaking change planned for the future for the ppx, this PR should actually *help* @rickyvetter and any other authors build and release that change, because once there are more tests, we will know exactly what cases are breaking after the change.

In other words, if this PR gets merged, nobody should notice anything, except the ppx maintainers and contributors, who would get a quality of life improvement thanks to the ppx now being part of the testing infrastructure in this repo.

jchavarri commented:
Oh, I think I see what you mean. If PPX lives in compiler repo, then it's easy to configure `bsconfig` flags like it was done with jsx2 and 3. 😞 I didn't think of that.

Maybe the parser could offer a way to switch between syntaxes per file with an attribute, like Reason parser is doing: https://github.com/facebook/reason/pull/2605/files#diff-2f961a549953a72af3a11bb3b36ecc49R4975.

This would be helpful not only for ppx of course, but any breaking syntax change.

jchavarri commented:
@chenglou I thought more about this. Because `syntax` is consumed as a git submodule inside compiler repo, nothing would prevent having 2 ppx modules with 2 different versions, and then adding any `bsconfig` configuration in compiler repo like it's been done in the past. From the compiler perspective, nothing changes either after this change, except it has to read the ppx file from another location. I'm curious to know what @bobzhang thinks as well.

bobzhang commented:
@jchavarri Ricky (@rickyvetter) owns the module.
There's a  concern since we still need support the reason syntax, so if we spins off the jsx ppx, how would it affect the old reason syntax

jchavarri commented:
@bobzhang There should be no impact in old Reason syntax. From the compiler perspective, this PR can be considered for all effects and purposes like moving a file from one folder to another.

bobzhang commented:
As long as it does not make the main maintainer’s life more difficult, I am
fine with it.
But keep in mind this module is used by reason syntax too
Javier Chávarri <notifications@github.com>于2020年8月31日 周一下午9:03写道：

>
>
> @bobzhang <https://github.com/bobzhang> There should be no impact in old
> Reason syntax. From the compiler perspective, this PR can be considered for
> all effects and purposes like moving a file from one folder to another.
>
>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rescript-lang/syntax/pull/124#issuecomment-683763959>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFWMK52VPL4VL46JZEICSDSDONTHANCNFSM4QPX2FPQ>
> .
>
>
> --
Regards
-- Hongbo Zhang

jchavarri commented:
I created a draft PR to show how minimal the impact downstream for the compiler is: https://github.com/rescript-lang/rescript-compiler/pull/4660.

rickyvetter commented:
I think this move could make maintaining quite a bit cleaner but I want to clarify setup. In the near future we need to make a series of major breaking changes to the ppx and need to ship both versions to the compiler so that folks can update their ppx version independently of their compiler version. Would you propose just exposing both versions as different modules here and then doing the switch logic in the compiler based on config?

jchavarri commented:
Hey @rickyvetter ! 👋 

> Would you propose just exposing both versions as different modules here and then doing the switch logic in the compiler based on config?

Yes, exactly. The [companion PR](https://github.com/rescript-lang/rescript-compiler/pull/4660/) in the compiler repo shows how the ppx module is available just like before (see `ppx_entry` module). I only had to change the linking order so that `napkin.cmxa` lib (this repo) is linked before the `syntax.cmxa` lib (that contains the ppx entry in compiler), which depends on it.

If there is a new version of the ppx, there can be another module added next to the one in this PR. Keeping these modules close to the testing infra might be beneficial too to guarantee the migration is smooth, and new tests can also be added from user reports. It seems the compiler [still keeps](https://github.com/rescript-lang/rescript-compiler/pull/4660/files#diff-41bdcb78363b25f759469b2591ccd0e2R31) the ppx version handling from bsconfig option.

The only thing that will change from previous setup is that every time there is a fix or update in the ppx, the compiler repo will have to sync the submodule version, while before the file could be just directly updated in-repo. I hope this is not a deal breaker.

Maybe I'm missing something :) Please let me know if there's anything else that I'm not accounting for 👍

rickyvetter commented:
Gotcha. Seems like a win to me. @chenglou you seemed hesitant at first - okay with this plan?

jchavarri commented:
updated the PR with the following:
- merge the changes from `master`
- add flag `-ppx` to the main `rescript.exe` binary (as discussed with @IwanKaramazow offline)
- remove ppx-related stuff from multiprinter (as jsx ppx seems very unrelated to printing really)
- skip roundtrip tests for specs that have `ppx` defined
- added some test `externalWithCustomName` that was failing before but now passes, I guess some recent change fixed it.

lmk what you think :)

IwanKaramazow commented on src/res_cli.ml:
```diff
@@ -243,8 +251,12 @@ module CliArgProcessor = struct
           else exit 1
         end
         else
+          let parsetree = match ppx with
```
Can we keep this a string and match on the string for now? Does this extra conversion give us something in the end?

IwanKaramazow commented on Makefile:
```diff
@@ -49,6 +50,8 @@ lib/rescript.exe: $(CLI_FILES)
 
 build-native: lib/refmt.exe lib/rescript.exe depend
 
+build-ppx: src/reactjs_jsx_ppx.cmx
```
why does the ppx need a separate `build-ppx` rule? If it's a part  "API_FILES", is this sufficient?

IwanKaramazow commented on src/res_cli.ml:
```diff
@@ -178,6 +179,7 @@ end = struct
   let origin = ref ""
   let interface = ref false
   let report = ref "pretty"
+  let ppx = ref "none"
```
can we default this to the empty string `""`?

IwanKaramazow commented on src/res_cli.ml:
```diff
@@ -227,6 +230,11 @@ module CliArgProcessor = struct
       | _ -> false
       in
 
+      let ppx = match ppx with
```
See comment down, this conversion doesn't add that much value?

IwanKaramazow commented on tests/runner.js:
```diff
@@ -204,15 +202,16 @@ let makeReproducibleFilename = (txt) => {
 };
 
-global.runPrinter = (dirname) => {
+global.runPrinter = (dirname, ppx = "none") => {
```
Can we make this the empty string or just keep it undefined as default value?

IwanKaramazow commented on tests/runner.js:
```diff
@@ -226,7 +225,7 @@ Make sure the test input is syntactically valid.`;
       }
 
-      if (process.env.ROUNDTRIP_TEST) {
+      if (process.env.ROUNDTRIP_TEST && ppx === "none") {
```
Can you add a comment here on why we don't run ppx'es in the roundtrip tests?

IwanKaramazow commented on tests/printer/reactJsx/__snapshots__/render.spec.js.snap:
```diff
@@ -0,0 +1,101 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
I wonder if we should make a separate folder for ppx tests: `tests/ppx/...`? What do you think?

IwanKaramazow commented on src/res_cli.ml:
```diff
@@ -258,19 +270,23 @@ module CliArgProcessor = struct
       exit 1
     | _ -> exit 1
-  [@@raises exit]
+  [@@raises Invalid_argument, exit]
```
What raises `Invalid_argument`?

jchavarri replied:
The existing react ppx uses that exception to note cases with unexpected situations. I can count at least 19 different places, e.g.

https://github.com/jchavarri/syntax/blob/23336db4e0b7d9f524db573b4e94afe510db43ab/src/reactjs_jsx_ppx.ml#L351-L354

IwanKaramazow reviewed (commented):
Left a couple of questions, what do you think?

jchavarri commented:
@IwanKaramazow I think I tackled all the comments, let me know if there's anything missing :)

IwanKaramazow reviewed (approved):
@jchavarri Looking good, thank you for all the work.

bobzhang commented:
This is not finished since the removal of jsx ppx is not done in the compiler.

After some thoughts, I think this may make your test workflow more difficult, @rickyvetter , since now you make changes to the ppx, you need update the syntax submodule to test your new changes

IwanKaramazow commented:
@bobzhang I'll take care of it.

jchavarri commented:
@bobzhang I had created a companion PR in https://github.com/rescript-lang/rescript-compiler/pull/4660/ that included the compiler changes. I didn't continue working on this because @chenglou mentioned on DM a few days after this PR was merged that the move to the PPX to the parser might be rethought / retracted because of the new jsx transform.

IwanKaramazow commented:
Let's sync up on this first.

jchavarri commented:
I talked to @IwanKaramazow and these are the steps to take before this is finished:

1. Update `syntax` repo so that `Reactjs_jsx_ppx` becomes `Reactjs_jsx_ppx_v3` again, as it will be useful later on, when new JSX ppx versions will be added.
2. Update https://github.com/rescript-lang/rescript-compiler/pull/4660 to with `master` and latest `syntax`, and make sure everything compiles.
